packages/engine/src/scene/components/VolumetricComponent.ts

```diff
@@ -132,200 +136,120 @@ export function VolumetricReactor() {
-          }
-        })
+      audioNodes.gain.gain.setTargetAtTime(volumetric.volume.value, audioContext.currentTime, 0.1)
     }
   }, [])
 
-  useEffect(
-    function updateVolume() {
-      const volume = volumetric.volume.value
-      const element = getOptionalComponent(entity, MediaElementComponent)?.element as HTMLMediaElement
-      if (!element) return
-      const audioNodes = AudioNodeGroups.get(element)
-      if (audioNodes) {
-        audioNodes.gain.gain.setTargetAtTime(volume, audioContext.currentTime, 0.1)
-      }
-    },
-    [volumetric.volume, volumetric.player]
-  )
-
   useEffect(() => {
-    if (!volumetric.player.value) return
-    if (volumetric.hasTrackStopped.value) {
-      // Track is changed. Set the track path
-      if (volumetric.paths[volumetric.track.value].value) {
-        volumetric.loadingEffectActive.set(volumetric.useLoadingEffect.value) // set to user's value
-        volumetric.loadingEffectTime.set(0)
-        volumetric.player.value.setTrackPath(volumetric.paths[volumetric.track.value].value)
-        volumetric.hasTrackStopped.set(false)
-      }
-    } else {
-      /** Track isn't changed. Probably new path is added or edited.
-       * No need to set track path.
-       */
-    }
-  }, [volumetric.track, volumetric.paths, volumetric.player, volumetric.hasTrackStopped])
-
-  useEffect(() => {
-    if (!volumetric.player.value) return
-    if (volumetric.paused.value) {
-      volumetric.player.value.pause()
-    } else {
-      volumetric.player.value.play()
+    const pathCount = volumetric.paths.value.length
+    if (!volumetric.ended.value || pathCount === 0) {
+      return
     }
-  }, [volumetric.paused, volumetric.player])
 
-  return null
-}
-
-export const endLoadingEffect = (entity, object) => {
-  if (!hasComponent(entity, AvatarEffectComponent)) return
-  const plateComponent = getComponent(entity, AvatarEffectComponent)
-  plateComponent.originMaterials.forEach(({ id, material }) => {
-    object.traverse((obj) => {
-      if (obj.uuid === id) {
-        obj['material'] = material
+    /**
+     * Find the next valid track: If the current track is invalid, try the next one.
+     * Here invalid means, path does not end with '.manifest'.
+     */
+    let currentTrack = volumetric.track.value
+    // eslint-disable-next-line
+    while (true) {
+      const nextTrack = getNextTrack(volumetric.playMode.value, pathCount, currentTrack)
+      const manifestPath = volumetric.paths.value[nextTrack]
+      if (
+        manifestPath &&
+        (manifestPath.endsWith('.manifest') || manifestPath.endsWith('.mp4') || manifestPath.endsWith('.json'))
+      ) {
+        currentTrack = nextTrack
+        break
+      } else if (nextTrack == volumetric.track.value) {
+        // No valid tracks found, stop playing
+        currentTrack = -1
+        break
+      } else {
+        // 'nextTrack' is not a valid track, try the next one
```

Comment:
What if all the tracks are invalid?

Reply:
It is handled in the below condition (L187). If all tracks are invalid, nothing is played, until there is any change in paths. I think I can clean up this while loop into a cleaner recursion. I'll do that.

Reply:
I realized while loop is unavoidable. Unless we iterate all tracks, we cannot know if all tracks are invalid.